File: src/main/java/gargoyle/xenox/util/res/cache/Cache.java
package gargoyle.xenox.util.res.cache;

import java.util.Collections;
import java.util.Map;
import java.util.WeakHashMap;

public class Cache {
    public static final Cache GLOBAL = new Cache();
    private final Map<String, Object> cache = Collections.synchronizedMap(new WeakHashMap<>());

    @SuppressWarnings("unchecked")
    public <R> R get(String location) {
        return (R) cache.get(location);
    }

    public <R> void put(String location, R resource) {
        cache.put(location, resource);
    }
}

File: src/main/kotlin/gargoyle/xenox/util/res/cache/Cache.kt
package gargoyle.xenox.util.res.cache

import java.util.Collections
import java.util.WeakHashMap

class Cache {
    private val cache = Collections.synchronizedMap(WeakHashMap<String, Any>())
    operator fun <R> get(location: String): R? {
        @Suppress("UNCHECKED_CAST")
        return cache[location] as R?
    }

    fun <R> put(location: String, resource: R) {
        cache[location] = resource
    }

    companion object {
        val GLOBAL = Cache()
    }
}
